ReactNativeClient/lib/reducer.js

```diff
@@ -337,27 +356,27 @@ function changeSelectedNotes(state, action, options = null) {
 		newState.selectedNoteIds = noteIds;
 		newState.selectedNoteHash = action.hash ? action.hash : '';
 
-		const backwardHistoryNotes = newState.backwardHistoryNotes.slice();
+		let backwardHistoryNotes = newState.backwardHistoryNotes.slice();
 		let forwardHistoryNotes = newState.forwardHistoryNotes.slice();
 
 		// The historyAction property is only used for user-initiated actions and tells how
 		// the history stack should be handled. That property should not be present for
 		// programmatic navigation. Possible values are:
 		// - "goto": When going to a note, but not via the back/forward arrows.
-		// - "pop": When clicking on the Back arrow
-		// - "push": When clicking on the Forward arrow
+		// - "goBackward": When clicking on the Back arrow
+		// - "goForward": When clicking on the Forward arrow
 		const lastSeenNote = stateUtils.getLastSeenNote(state);
 		if (action.historyAction == 'goto' && lastSeenNote != null &&  action.id != lastSeenNote.id) {
 			forwardHistoryNotes = [];
-			backwardHistoryNotes.push(Object.assign({}, lastSeenNote));
-		} else if (action.historyAction === 'pop' && lastSeenNote != null) {
+			backwardHistoryNotes = backwardHistoryNotes.concat(lastSeenNote).slice(-MAX_HISTORY);
+		} else if (action.historyAction === 'goBackward' && lastSeenNote != null) {
 			if (forwardHistoryNotes.length === 0 || lastSeenNote.id != forwardHistoryNotes[forwardHistoryNotes.length - 1].id) {
-				forwardHistoryNotes.push(Object.assign({}, lastSeenNote));
+				forwardHistoryNotes = forwardHistoryNotes.concat(lastSeenNote).slice(-MAX_HISTORY);
```

Comment:
It's what happens in browsers when you click a link. Also in file explorers.

Reply:
Is not this code for when I go backwards?  I was confused by the slice, which I suspect is redundant here.  Or have I misunderstood?

Reply:
Yes, I guess it is redundant since forwardHistoryNotes is dependent on how many times we can go backward, which is limited by the slice of backwardHistoryNotes.